{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f85ccd1a",
   "metadata": {},
   "source": [
    "### Using BeautifulSoup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e74eb60e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "from urllib.request import urlopen\n",
    "from tqdm import tqdm\n",
    "from urllib.parse import urljoin\n",
    "\n",
    "def get_soup(url):\n",
    "    page = urlopen(url)\n",
    "    html = page.read().decode(\"utf-8\")\n",
    "    return BeautifulSoup(html, \"html.parser\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2d305095",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/560491 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/560491 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/536292 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/536292 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/528834 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/528834 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/518339 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/518339 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/176670 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/176670 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/159661 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/159661 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/138588 , due to:  list index out of range\n",
      "Could not retreive pdf from:  https://katalogoa.siis.net/Record/138588 , due to:  list index out of range\n",
      "Wall time: 4min 1s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Getting the url to the content subpages\n",
    "\n",
    "base_url = 'https://katalogoa.siis.net'\n",
    "url = \"https://katalogoa.siis.net/Search/Results?lookfor=LFCL&type=AllFields&sort=year&page=1&filter%5B%5D=quitar_facet%3A%22database_facet%22\"\n",
    "soup = get_soup(url)\n",
    "\n",
    "i = 1\n",
    "content_file_urls = []\n",
    "\n",
    "while len(soup.find_all('a', class_=\"title\"))>0:\n",
    "    \n",
    "    # Get the html content\n",
    "    url = \"https://katalogoa.siis.net/Search/Results?lookfor=LFCL&type=AllFields&sort=year&page=\" + str(i) + \"&filter%5B%5D=quitar_facet%3A%22database_facet%22\"\n",
    "    soup = get_soup(url)\n",
    "\n",
    "    # Parse the url with beautifulsoup\n",
    "    for card in soup.find_all('a', class_=\"title\"):\n",
    "        content_url = card['href']\n",
    "        if not content_url.startswith('http'):\n",
    "            content_url = urljoin(base_url, content_url)\n",
    "\n",
    "        content_soup = get_soup(content_url)\n",
    "        try:\n",
    "            content_file_urls.append(content_soup.find_all('a', class_=\"fulltext boton-descargar\")[0]['href'])\n",
    "        except Exception as e:\n",
    "            print(\"Could not retreive pdf from: \", content_url, \", due to: \", e)\n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0c147edc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "201"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# removing duplicates\n",
    "content_file_urls = list(dict.fromkeys(content_file_urls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "35a7da3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████| 201/201 [00:00<00:00, 200966.65it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "# Saving files list in memory\n",
    "try:\n",
    "    if content_file_urls:\n",
    "        with open(r'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\content_list.txt', 'w') as fp:\n",
    "            for item in tqdm(content_file_urls):\n",
    "                # write each item on a new line\n",
    "                fp.write(\"%s\\n\" % item)\n",
    "except Exception as e:\n",
    "    print(e)\n",
    "        \n",
    "# Reading files list from memory\n",
    "try:\n",
    "    with open('C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\content_list.txt') as f:\n",
    "        content_file_urls = f.read().splitlines()\n",
    "except Exception as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "987df0d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "178\n"
     ]
    }
   ],
   "source": [
    "# 178 out of the 208 pages with 'lectura fácil' contains usable .pdf files\n",
    "\n",
    "counter_pdf = 0\n",
    "\n",
    "for url in content_file_urls:\n",
    "    if '.pdf' in url:\n",
    "        counter_pdf += 1\n",
    "\n",
    "print(counter_pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ff580072",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking in indexes: https://pypi.org/simple, https://OelQQDud:****@alm.accenture.com/nexus/repository/ISCP_Components_Release/simple/\n",
      "Requirement already satisfied: pip-system-certs in c:\\users\\fernando.rubio.perez\\anaconda3\\lib\\site-packages (4.0)\n",
      "Requirement already satisfied: wrapt>=1.10.4 in c:\\users\\fernando.rubio.perez\\anaconda3\\lib\\site-packages (from pip-system-certs) (1.14.1)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|▊                                                                                 | 2/201 [00:04<06:48,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Vidas_de_mujeres_inspiradoras_4.pdf?sequence=1&rd=003198715430455'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5%|████                                                                             | 10/201 [00:16<04:51,  1.53s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 12%|█████████▋                                                                       | 24/201 [00:25<01:36,  1.83it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\default.aspx?CodigoCompleto=Observatoriors@@@CAS_ley_foral_infancia.pdf'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 19%|███████████████▋                                                                 | 39/201 [00:42<02:52,  1.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Vidas_de_mujeres_inspiradoras_3.pdf?sequence=1&rd=003181315126266'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 20%|████████████████▌                                                                | 41/201 [00:43<02:16,  1.17it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 24%|███████████████████▋                                                             | 49/201 [00:51<01:45,  1.44it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\2ad695d3-39d3-42b7-4af9-853c496fc5c8?t=1657633844058'\n",
      "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 28%|██████████████████████▉                                                          | 57/201 [00:56<01:43,  1.38it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 44%|███████████████████████████████████▍                                             | 88/201 [01:11<00:58,  1.93it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Vidas_de_mujeres_inspiradoras_en_lectura_f%c3%a1cil.pdf?sequence=1&rd=0031323571172084'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 47%|█████████████████████████████████████▉                                           | 94/201 [01:19<02:35,  1.45s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 57%|█████████████████████████████████████████████▊                                  | 115/201 [02:07<02:38,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 67%|█████████████████████████████████████████████████████▋                          | 135/201 [02:31<00:59,  1.10it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HTTPSConnectionPool(host='sid.usal.es', port=443): Max retries exceeded with url: /idocs/F8/FDO27430/Guia_orientacion_juridicaLF.pdf (Caused by SSLError(SSLCertVerificationError(1, \"[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: Hostname mismatch, certificate is not valid for 'sid.usal.es'. (_ssl.c:1129)\")))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 70%|███████████████████████████████████████████████████████▋                        | 140/201 [02:40<01:33,  1.52s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Prevenci%c3%b3n_riesgos_laborales_Lavander%c3%ada_lectura_f%c3%a1cil.pdf?sequence=1&rd=0031541764220962'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      " 70%|████████████████████████████████████████████████████████                        | 141/201 [02:43<02:12,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Prevenci%c3%b3n_riesgos_laborales_Log%c3%adstica_lectura_f%c3%a1cil.pdf?sequence=1&rd=0031883942881233'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      " 71%|████████████████████████████████████████████████████████▌                       | 142/201 [02:44<01:51,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Prevenci%c3%b3n_riesgos_laborales_Limpieza_industrial_lectura_f%c3%a1cil.pdf?sequence=1&rd=0031199650012106'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      " 71%|████████████████████████████████████████████████████████▉                       | 143/201 [02:46<01:48,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Prevenci%c3%b3n_riesgos_laborales_Gu%c3%ada_general_lectura_f%c3%a1cil.pdf?sequence=1&rd=0031381834235355'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 80%|███████████████████████████████████████████████████████████████▋                | 160/201 [03:06<01:01,  1.51s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 90%|████████████████████████████████████████████████████████████████████████        | 181/201 [03:19<00:07,  2.55it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 94%|██████████████████████████████████████████████████████████████████████████▊     | 188/201 [03:28<00:13,  1.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Saber_mas_sobre_epilepsia.pdf?sequence=1&rd=0031118776874025'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 96%|████████████████████████████████████████████████████████████████████████████▍   | 192/201 [03:31<00:09,  1.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\La_capacidad_jur%c3%addica_y_la_tutela_LF.pdf?sequence=1&rd=0031260359806239'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 98%|██████████████████████████████████████████████████████████████████████████████  | 196/201 [03:34<00:04,  1.21it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 22] Invalid argument: 'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\\\\Convencion_Internacional_Naciones_Unidas_Derechos_Personas_Discapacidad.pdf?sequence=1&rd=0031445015439837'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████| 201/201 [03:36<00:00,  1.08s/it]\n"
     ]
    }
   ],
   "source": [
    "!pip install pip-system-certs\n",
    "import requests\n",
    "import os\n",
    "\n",
    "# Saving pdf content from urls into memory\n",
    "path_to_local_files = \"C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs\"\n",
    "pdfs_with_errors = []\n",
    "\n",
    "for url in tqdm(content_file_urls):\n",
    "    if '.pdf' in url:\n",
    "        try:\n",
    "            response = requests.get(url)\n",
    "            file_local_path = os.path.join(path_to_local_files, url.split('/')[-1])\n",
    "            with open(file_local_path, 'wb') as f:\n",
    "                f.write(response.content)\n",
    "\n",
    "        except Exception as e:\n",
    "            print(e)\n",
    "            pdfs_with_errors.append(url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5a71c296",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████| 21/21 [00:00<00:00, 20996.52it/s]\n"
     ]
    }
   ],
   "source": [
    "# 21 out of the 178 pdf files had some kind of problem when loading them\n",
    "\n",
    "print(len(pdfs_with_errors))\n",
    "\n",
    "with open(r'C:\\\\Users\\\\fernando.rubio.perez\\\\Jupyter projects\\\\TFM\\\\files_in_lectura_facil\\\\cedid\\\\pdfs_with_errors.txt', 'w') as fp:\n",
    "    for item in tqdm(pdfs_with_errors):\n",
    "        # write each item on a new line\n",
    "        fp.write(\"%s\\n\" % item)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
